### PR TITLE
Added developer docs to help onboard new contributors

### PR DESCRIPTION
**Pull request by tgaddair**

### CODE REVIEW

File: docs/developers.rst [hunk 1]
```diff
@@ -0,0 +1,173 @@
+
+.. inclusion-marker-start-do-not-remove
+
+
+Developer Guide
```
Comment (sblotner) [inline docs/developers.rst]: What do you think about calling it "Contribution Guide"?

File: docs/developers.rst [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+Environment Setup
+~~~~~~~~~~~~~~~~~
+
+Clone the repository locally:
+
+.. code-block:: bash
+
+    $ git clone --recursive https://github.com/horovod/horovod.git
+
+Be sure to run within a virtual environment to avoid dependency issues:
```
Comment (sblotner) [inline docs/developers.rst]: Maybe "Use a virtual environment to avoid dependency issues: "

File: docs/developers.rst [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+**IMPORTANT:** Some tests contain GPU-only codepaths that will be skipped if running without GPU support.
+
+
+Adding Custom Operations
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Operations in Horovod are used to transform Tensors across workers.  Horovod currently supports operations that
+implement Broadcast, Allreduce, and Allgather interfaces.  Gradients in Horovod are aggregated through
+Allreduce operations (with the exception of sparse gradients which use Allgather).
```
Comment (sblotner) [inline docs/developers.rst]: add comma after "gradients"

File: docs/developers.rst [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+All data transfer operations are implemented in the
+`horovod/common/ops <https://github.com/horovod/horovod/tree/master/horovod/common/ops>`__ directory.  Implementations
+are organized by the collective communication library used to perform the operation (e.g.,
+`mpi_operations.cc <https://github.com/horovod/horovod/blob/master/horovod/common/ops/mpi_operations.cc>`__ for MPI).
+
+To create a new custom operation, start by defining a new class that inherits from the base operation, in the file
+corresponding to the library you'll be using to implement the operation:
```
Comment (sblotner) [inline docs/developers.rst]: you'll use

File: docs/developers.rst [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+parameter provides access to all the Tensor buffers and metadata that need to be processed,
+and the ``response`` parameter contains additional metadata including which devices are being used by different ranks.
+
+``Enabled`` should return true if your operation can be performed on the given Tensor entries subject to the
+current parameter settings and response metadata.
+
+Once you've written the implementation for your operation, add it to the ``OperationManager`` in the
+``CreateOperationManager`` function of
+`operations.cc <https://github.com/horovod/horovod/blob/master/horovod/common/operations.cc>`__.  Because more than one
+operation may be *enabled* at a time, but only one will be performed on a given vector of Tensor entries, the order of
```
Comment (sblotner) [inline docs/developers.rst]: consider the order of ...

File: docs/developers_include.rst [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.. include:: ./development.rst
```
Comment (alsrgv) [inline docs/developers_include.rst]: Please fix the link

File: docs/developers.rst (hunk 1 above)
Comment (tgaddair) [reply docs/developers.rst]: Changed to "Contributor Guide".  What do you think?  (I prefer Contributor over Contribution as the former is more broad, meaning we can include stuff useful to contributors beyond explaining how to make contributions)

Comment (sblotner) [reply docs/developers.rst]: That works!

**Pull request merged**